File: legacy/matlab_source/ThomsonsMethodRevisitedExperiments/FastMultitaper.m
function FMTSE = FastMultitaper(N,W,cutoff,epsilon)
%% FastMultitaper.m 
% Function implementing the fast multitaper spectral estimator
% 
% Inputs:
% -N: Length of DPSSs
% -W: Half Bandwidth for DPSSs
% -cutoff: Projection onto the first K = cutoff DPSSs 
%          OR Projection onto the first K = #{ell : lambda_{N,W}^{(ell)} >= cutoff} DPSSs.
% -epsilon: Relative error in approximation
%
% Outputs:
% -FMTSE an object containing a Fast Multitaper Spectral Estimate function
%
% Given an Nx1 vector x, performing
% y = FMTSE.SpectralEstimate(x,M);
% results in y[m] \approx (1/K)x^*E_(m/M)S_KS_K^*E_(m/M)^*x
% i.e. y[m] is approximately the multitaper spectral estimate at f = m/M
%
% Notes: 
%
% Most recent change - 01/30/2019
%
% Copyright 2019, Santhosh Karnik

%% Fast Multitaper Function
FMTSE.SpectralEstimate = @SpectralEstimate;

%% Store signal length, bandwidth, tolerance, and number of DPSSs
FMTSE.N = N; % Signal Length
FMTSE.W = W; % Half-Bandwidth
FMTSE.cutoff = cutoff; % Number of DPSSs
FMTSE.epsilon = epsilon; % Tolerance

%% Compute DPSSs and store transition region DPSSs
[S,lambda,lambda_all,lowerindex,upperindex] = transitionDPSS(N,W,epsilon,cutoff);
if(cutoff >= 1)
    K = cutoff;
    eig_weights = [ones(K-lowerindex+1,1); zeros(upperindex-K,1)] - lambda;
elseif(cutoff > eps)
    eig_weights = (lambda > cutoff) - lambda; %positive and negative weighing of weights to eigs
    K = nnz(lambda > cutoff) + lowerindex-1;
end

FMTSE.S = bsxfun(@times,S,sqrt(abs(eig_weights'))); %the abolsute of the eigen weigh takes care of the negative values
FMTSE.index_plus = (eig_weights > 0);
FMTSE.r = length(eig_weights);
FMTSE.K = K;
FMTSE.lambda = lambda;
FMTSE.lambda_all=lambda_all;

%% Precompute entries of the prolate matrix B_{N,W}.
FMTSE.vecHalfSinc = 2*W*sinc(2*W*(1:(N-1))');


%% Fast Multitaper Spectral Estimation
function z = SpectralEstimate(x,M)
    if(FMTSE.N ~= size(x,1))
        disp('ERROR: x has incorrect length')
    elseif(size(x,2) > 1)
        disp('x must be a single column vector')
    else        
        % Compute eigenvalue-weighted multitaper spectral estimate using FFTs
        if(nargin < 2)
            M = FMTSE.N; 
        end
        L = M*ceil(2*FMTSE.N/M);
        vecSinc = [2*FMTSE.W;FMTSE.vecHalfSinc;zeros(L-2*FMTSE.N+1,1);flipud(FMTSE.vecHalfSinc)];
        Fx = fft(x,L,1);
        z0 = ifft(fft(real(Fx).^2+imag(Fx).^2,L,1).*vecSinc,L,1);
        if(L > M)
            z0 = z0(1:(L/M):end);
        end

        % Compute tapered spectral estimates using transition-region tapers
        if(FMTSE.N > M)
           Sx = bsxfun(@times,FMTSE.S,x);
           % Datawrap Sx
           A = zeros(M,FMTSE.r);
           for k = 1:FMTSE.r
               A(:,k) = datawrap(Sx(:,k),M);
           end
        else
            A = bsxfun(@times,FMTSE.S,x);
        end
        A = fft(A,M,1);
        A = A.*conj(A); %real(A).^2+imag(A).^2;
        z1 = sum(A(:,FMTSE.index_plus),2)-sum(A(:,~FMTSE.index_plus),2);
        
        % Add eigenvalue-weighted estimate to transition-region correction
        z = (z0+z1)/FMTSE.K;
        
        % Set any entries <=0 to machine precision times maximum
        maxz = max(z);
        z = max(z,eps*maxz);
    end
end

end

File: python/thomson_multitaper/fast_multitaper.py
"""Translation of FastMultitaper.m: FFT-accelerated multitaper spectral
estimator using transition-region DPSS correction. Original Copyright 2019,
Santhosh Karnik.
"""
from __future__ import annotations

import numpy as np

from ._dpss_transition import transition_dpss
from ._utils import datawrap


class FastMultitaper:
    def __init__(self, n: int, w: float, cutoff: float, epsilon=None):
        self.n = n
        self.w = w
        self.cutoff = cutoff
        self.epsilon = epsilon
        eps = np.finfo(float).eps

        s, lam, lambda_all, lowerindex, upperindex = transition_dpss(n, w, epsilon, cutoff)

        if cutoff >= 1:
            k = cutoff
            eig_weights = np.concatenate([np.ones(k - lowerindex + 1), np.zeros(upperindex - k)]) - lam
        elif cutoff > eps:
            eig_weights = (lam > cutoff).astype(float) - lam
            k = int(np.count_nonzero(lam > cutoff)) + lowerindex - 1
        else:
            raise ValueError("cutoff must be >= 1 or > eps")

        # Defensive guard, not part of the original translation: at NW below ~3 (this cutoff),
        # k=0 -- no taper reaches the `cutoff` eigenvalue threshold at all. Every downstream use
        # of self.K divides by it (spectral_estimate here, fast_spectrum_batch in
        # ccf_pipeline/fast_cross_spectrum.py), silently producing NaN/Inf rather than a clear
        # error -- found live during Round 2's NW sweep (2026-09-04), where 65/300 low-bandwidth
        # sweep points failed with a confusing "need at least one array to concatenate" deep
        # inside the picker instead of a diagnosable cause. Raising here changes nothing for any
        # K>0 input (every existing call site already assumed K>0); it only replaces silent
        # corruption with an informative failure at the one input range that was never valid.
        if k == 0:
            raise ValueError(
                f"FastMultitaper: NW={n * w:.3g} (cutoff={cutoff}) gives K=0 -- no taper reaches "
                f"the cutoff eigenvalue threshold. NW must be large enough that at least one "
                f"taper's eigenvalue exceeds `cutoff`; empirically this project's cutoff=1-1e-5 "
                f"needs NW >~ 3."
            )

        self.S = s * np.sqrt(np.abs(eig_weights))[None, :]
        self.index_plus = eig_weights > 0
        self.r = len(eig_weights)
        self.K = k
        self.lambda_ = lam
        self.lambda_all = lambda_all

        self.vec_half_sinc = 2 * w * np.sinc(2 * w * np.arange(1, n))

    def spectral_estimate(self, x: np.ndarray, m: int | None = None) -> np.ndarray:
        if self.n != x.shape[0]:
            raise ValueError("x has incorrect length")
        if x.ndim > 1 and x.shape[1] > 1:
            raise ValueError("x must be a single column vector")

        if m is None:
            m = self.n
        eps = np.finfo(float).eps

        l = m * int(np.ceil(2 * self.n / m))
        vec_sinc = np.concatenate(
            [
                [2 * self.w],
                self.vec_half_sinc,
                np.zeros(l - 2 * self.n + 1),
                self.vec_half_sinc[::-1],
            ]
        )
        fx = np.fft.fft(x, n=l, axis=0)
        z0 = np.fft.ifft(np.fft.fft(fx.real**2 + fx.imag**2, n=l, axis=0) * vec_sinc, n=l, axis=0)
        if l > m:
            z0 = z0[0 :: l // m]

        if self.n > m:
            sx = self.S * x[:, None]
            a = np.stack([datawrap(sx[:, k], m) for k in range(self.r)], axis=1)
        else:
            a = self.S * x[:, None]

        a = np.fft.fft(a, n=m, axis=0)
        a = a * np.conj(a)
        z1 = a[:, self.index_plus].sum(axis=1) - a[:, ~self.index_plus].sum(axis=1)

        z = (z0 + z1) / self.K
        z = z.real  # z0/z1 are real up to FFT round-off; matches MATLAB's implicit real cast in max()
        max_z = z.max()
        return np.maximum(z, eps * max_z)
